CIPs/0009.md

```diff
@@ -0,0 +1,86 @@
+
+Option `1` results in downtime, option `2` risks double signing, and option `3` can only be done once every 24 hours.
+Furthermore, options `1` and `2` may require unsafe key management practices, which may result in keys being compromised.
+
+Decoupling key management from other validator functionality will allow validators to switch nodes with zero downtime and no risk of double signing.
+
+## Goals
+
+- Zero downtime validator upgrades within an epoch
+- Failover possible with no risk of double signing
+
+## Proposed Solution
+
+The proposed solution is to use `clef`, introduced in [go-ethereum v1.9](https://blog.ethereum.org/2019/07/10/geth-v1-9-0/), as a standalone signer for validators.
+
+`clef` is integrated into `celo-blockchain` as an implementation of the [wallet](https://github.com/celo-org/celo-blockchain/blob/3caca596d1ff23a9faf5339f108c993f38bfa743/accounts/accounts.go#L50) interface in [accounts/external/backend.go](https://github.com/celo-org/celo-blockchain/blob/master/accounts/external/backend.go).
+
+Validators will run a `clef` instance that manages their [validator signer key](https://docs.celo.org/validator-guide/summary/detailed#authorized-validator-signers), and point their node to that instance.
+
+To switch nodes, validators will issue a command to `clef` to stop accepting RPCs from node A and start accepting RPCs from node B.
+
+### Implementation
+
+#### Wallet interface
+The current `clef` implementation of the `wallet` interface is missing the following functions needed to support running a validator:
+- Decrypt
+- SignHashBLS
+- SignMessageBLS
+- GenerateProofOfPossession
+- GenerateProofOfPossessionBLS
+- GetPublicKey
+
+To implement these functions, new RPCs will need to be added to `clef`.
+
+#### Node mutex
+More investigation will need to be done to figure out exactly how to atomically switch nodes.
+
+Some options:
+
+- Implement a `clef` RPC to switch the value of the `--rpcaddr` flag.
+- Have `clef` accept RPCs from both nodes, and use `miner.start` and `miner.stop` on the node (or similar RPCs that allow these operations to be scheduled)
+
+#### Threshold signatures
+Using threshold signatures for block finalization is an approach taken by [Polychain Labs' Cosmos validator](https://blog.polychainlabs.com/tendermint/2020/03/26/threshold-validator-for-tendermint.html) to improve security and availability.
+
+Celo is a natural candidate to adopt a similar feature for BLS block finalization signatures and ECDSA consensus message signatures.
+
+A standalone signing service would simplify the implementation of this feature.
+
+Validators could run `N` instances of `clef`, and one or more instances of `celo-blockchain`.
+
+The `celo-blockchain` nodes could be configured to ping each of the `clef` instances for a signature share, and return once `M` have been collected.
+
+The `M/N` threshold signature would protect against `N-M` `clef` instances being unavailable or compromised.
```

Comment:
If the ECDSA key lives on a single machine, then an adversary A only has to compromise that single machine to authorize a new BLS public key of A's choice. To achieve `M/N` threshold security for signing, threshold ECDSA signing would need to be implemented as well.

Note BLS public keys can only be updated once a day, so compromise of the ECDSA key is not immediately equivalent to compromise of the BLS key (although within 24 hours it is).

Reply:
That's a good point. It would be better to have an ECDSA consensus key that cannot change BLS keys.

Reply:
What if the last BLS key has to sign off on the new one? As long as the initial BLS public key, which is still signed by the ECDSA public key, is the same as the one resulting from the DKG, then (m, n)-threshold security should hold for future key rotation updates.

There may be other good reasons to implement ECDSA threshold signatures, but if we can get the same security properties by relying just on our threshold BLS implementation, that keeps our attack surface smaller.

Reply:
I think the best option here is to have an ECDSA key that is *only* allowed to sign consensus messages, as @kobigurk suggested

Reply:
Added in a *future work* section below

Reply:
Or to move to only use BLS for consensus messages..